consensus/src/ledger/indexed_merkle_tree.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{marker::PhantomData, sync::Arc};
+
+use anyhow::*;
+use indexmap::IndexSet;
+use snarkos_storage::Digest;
+use snarkvm_algorithms::{merkle_tree::MerkleTree, MerkleParameters};
+use snarkvm_utilities::ToBytes;
+
+pub struct IndexedMerkleTree<P: MerkleParameters> {
+    tree: MerkleTree<P>,
+    index: IndexSet<Digest>,
+}
+
+impl<P: MerkleParameters> Clone for IndexedMerkleTree<P> {
+    fn clone(&self) -> Self {
+        let tree = self
+            .tree
+            .rebuild::<[u8; 32], _>(DummyLength(self.index.len(), PhantomData), &[])
+            .expect("failed to clone merkle tree");
+        Self {
+            tree,
+            index: self.index.clone(),
+        }
+    }
+}
+
+struct DummyLength<T>(usize, PhantomData<T>);
```

Comment:
I'd rename it to be more clear, e.g. `DummyCommitmentCount` or something similar

Reply:
its generically a dummylength. doesn't need to have any coupling to commitments. just a BS exactsizeiterator until later snarkvm PRs land

Reply:
I'm ok with this if it's just a temporary placeholder, I'd just make a comment on this.